### ISSUE TITLE
[32] Florence: A New Foundation Model for Computer Vision

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#60]
[paper](https://arxiv.org/abs/2111.11432)

vision-language contrastive pretraining 이후 downstream task에 transfer. 온갖 벤치마크에서 sota를 찍음

# Abstract
3가지를 강조한다
1. from coarse (scene) to fine (object)
2. from static (images) to dynamic (videos)
3. from RGB to multiple modalities (caption, depth)

![image](https://user-images.githubusercontent.com/16400591/142985900-d73ed977-865a-44a6-994d-f956a873acca.png)

# Intro
최근 microsoft 에서는 [XYZ-code](https://www.microsoft.com/en-us/research/blog/a-holistic-representation-toward-integrative-ai/) google에서는 [pathways](https://blog.google/technology/ai/introducing-pathways-next-generation-ai-architecture/) 를 공개했다. 모두 single model 로 여러가지 task 를 풀어내는 방법들이다.

이들을 foundation model 이라 칭하겠다. foundation model 이라는 용어는 [On the Opportunities and Risks of Foundation Models
](https://arxiv.org/abs/2108.07258) 에서 처음 소개되었다. CLIP, ALIGN, Wu Dao 2.0 도 foundation model 로 최근에 나왔다. 

![image](https://user-images.githubusercontent.com/16400591/142987049-32948079-30b6-48a5-844f-cd7dc9fa4601.png)
Florence는 위 그림처럼 3가지 구성으로 이루어져 있다.
1. data curation
2. model pretraining
3. task adaptations and training infrascturue

# Approach

## Data Curation
말해, 뭐해. 데이터는 foundation model 의 피와 같다.
이를 위해 30억개의 image-text 데이터를 크롤링 했다. 
그리고 [ALIGN](https://arxiv.org/abs/2102.05918) 에서처럼 필터링을 한다.
추가로 CLIP 이나, DALL-E 에서 사용하는 filtering 도 가져왔다.

Filter !!
1. 야한 이미지 제거
2. shorter dim < 200 제거
3. aspect ratio > 3 제거
4. 중복 이미지, 비슷한 이미지 제거
5. 1000개 token 이상의 설명이 달렸으면 제거
6. rare token 쓰는 text 제거 (100 million frequent 아닌 것)
7. 명백히 이미지와는 관련 없는 text 가 있으면 제거 (“1920x1080”, “alt img”, and “cristina”)
8. 너무 짧은 caption 제거
9. non-English 제거 (Python package cld3)
10. 쓰잘데기 없는 text 형식 제거 (“photographed on <date>” 등)

이렇게 하면 9억개가 남는다.
(FLD-900 set ==> FLorence Dataset)
9.7M 의 unique query 와 7.5B 의 token 을 갖는다.
900M 중에 9.7M 의 unique query가 있다는 소리는 description 이 동일한 이미지들이 꽤 많다는 소리 (1/100 확률로 같은 query)

## UniCL (Unified Image-Text Contrastive Learning)
크롤링 데이터가 노이즈가 많다 보니까 UniCL 을 썼다.
~UniCL 논문은 arxiv 준비중이라고 한다 ㅋㅋㅋ 어이없.. (Unified contrastive learning in image-text-label space.)~
그래도 간단하게는 설명해놨으니 살펴보자.

image-text 에서 triplet `(x, t, y)` 을 우선 생성한다.
x => image
t => language description
y => language label

짧은 description이 있다면 (1~2 word) 다음과 같이 prompt를 이용해서 enrich하게 text를 만든다.
```
“A photo of the [WORD]”
“A cropped photo of [WORD]”
```

### Notation
<img src="https://render.githubusercontent.com/render/math?math=f_\theta"> ==> image encoder (CoSwin transformer)
<img src="https://render.githubusercontent.com/render/math?math=f_\pi"> ==> text encoder (12 layer transformer)
u ==> normalized visual feature vector <img src="https://render.githubusercontent.com/render/math?math={f_\theta} \over {|f_\theta|}"> 
v ==> normalized language feature vector <img src="https://render.githubusercontent.com/render/math?math={f_\pi} \over {|f_\pi|}"> 
τ ==> learnable temperature
B ==> minibatch

### UniCL
2개의 contrastive term 을 쓴다.
같은 query 를 갖는 이미지는 positive 고, 나머지는 negative 로 활용한다.
batch를 어마무시하게 크게 사용하기 때문에 이런 방식이 먹히는 것으로 보인다 ㄷㄷㄷㄷ
CL 에서는 negative 숫자도 중요한데, 다 씹어먹는 크기인 24,576 배치를 쓴다.
그 중에 positive 는 섞여 있겠지 뭐~ 라는 마인드.
1/100 확률로 같은 query 라는 것은 이미 위에서 밝혀져 있으니... 그러려니 한다.

![image](https://user-images.githubusercontent.com/16400591/143824099-f55d4b42-7d3d-4bc2-a892-da7b801e359c.png)

supervised image to language contrastive loss
![image](https://user-images.githubusercontent.com/16400591/143824120-b544de65-6795-4535-be54-c1803e6cac98.png)
![image](https://user-images.githubusercontent.com/16400591/143824685-cab68b5e-ed97-4b6c-84a1-38c41ee1e13f.png)

supervised language to image contrastive loss
![image](https://user-images.githubusercontent.com/16400591/143824158-b71d97f7-1172-42e8-97d7-33945bf0261e.png)
![image](https://user-images.githubusercontent.com/16400591/143824707-fd6a3cd9-5f08-4efe-8abf-2efd41bf5f49.png)

위 수식을 조금 풀어 보자면, 같은 batch 내에서 같은 language 끼리 contrastive loss 를 적용하고 평균을 내는 것이다.
learnable τ 는 smoothing 역할을 해주는 것으로 보인다.

## Details
짧은 description은 prompt 를 붙여서 사용한다고 했는데, 그게 정답이 아닐 수 있다. 모델에 악영향이 갈 수 있어서 2-step training 을 수행한다.
1. augmented data 포함해서 1M iteration 돌림
2. augmented data 전부 제외하고 180K iteration 돌림

Adam + decoupled weight decay
224x224
max language description 76 token
batch size 24,576 (?!?!)
classification finetune 384x384, 80K iteration
512 A100 으로 10일동안 pretrain

## Model
image encoder 로 CoSwin transformer 를 쓴다.
swin 에서 patch embedding 이랑 patch merging 부분을 CvT 에서 제안했던 convolutional embedding 으로 두는 것이다.

text encoder 로는 12 layer transformer

## Object-level Visual Representation Learning
pretrained image encoder (CoSwin) 에다가 [Dynamic Head](https://openaccess.thecvf.com/content/ICCV2021/papers/Dai_Dynamic_DETR_End-to-End_Object_Detection_With_Dynamic_Attention_ICCV_2021_paper.pdf) 만 추가했다.
[one-stage ATSS framework](https://arxiv.org/abs/1912.02424v4) 으로 학습.

FLOD-9M (for FLorence Object detection Dataset) 을 구축했다. (OD 용 pretrain 용도)
COCO + LVIS + OpenImages + Object365 + pseudo labeled ImageNet-22K
pseudo label 방식은 [Rethinking Pre-training and Self-training](https://arxiv.org/pdf/2006.06882.pdf) 논문대로 했고, 내용은 다음과 같다. 
![image](https://user-images.githubusercontent.com/16400591/143830860-c8a4f2ab-be67-4914-bdb5-c62320d3655a.png)
FLOD-9M 은 8,967,286 images, 25,190 object categories, 33,408,237 bounding boxe 들을 포함한다.
이걸로 dynamic head 부분을 12epoch, 128 bs, 128 A100, 7일 학습했다.

![image](https://user-images.githubusercontent.com/16400591/142987088-50e6e36e-eddb-46d0-b681-73df37b335a4.png)

## Fine-Grained V+L Representation Learning
VQA, image captioning 이 해당됨.
[METER](https://arxiv.org/pdf/2111.02387.pdf) 에서 image encoder 부분만 CoSwin으로 대체했음
![image](https://user-images.githubusercontent.com/16400591/142987187-04580364-c50e-4786-a7d4-f90eff315178.png)

## Adaption to Video Recognition
[Video Swin-Transformer](https://arxiv.org/pdf/2106.13230.pdf) 처럼 viedo CoSwin 을 만들어냄
![image](https://user-images.githubusercontent.com/16400591/143856805-3af2f034-b769-4769-91ae-985ab9f435ca.png)

video coswin 은,
1. image tokenization layer 는 video tokenization layer 로 바꿈
2. 2d conv 의 weight 는 temporal dimension 으로 같은 값을 가지게 3d conv initialize
3. 2d patch merging 대신 3d patch merging 사용

## Scalable Training Infrastructure
1. Zero Redundancy Optimizer (ZeRO)
2. Activation Checkpointing
3. Mixed-precision Training
4. Gradient Cache

# Results

## Image-Text Retrieval
![image](https://user-images.githubusercontent.com/16400591/142987610-9e1bbb1d-e720-4568-93c8-bf4dd5a3ef6d.png)

![image](https://user-images.githubusercontent.com/16400591/142987542-158280c5-0454-4ee5-a67f-8a71d197683d.png)

![image](https://user-images.githubusercontent.com/16400591/142987503-12b103f9-79e1-44b6-bcf8-4e7299914aa5.png)

![image](https://user-images.githubusercontent.com/16400591/142987378-ab609ca8-7cc8-4d22-b336-f19200fe127d.png)
![image](https://user-images.githubusercontent.com/16400591/142987392-eb1e0775-e36c-4d33-82b6-66470ba7e419.png)
